win32/src/win32evtlog.i

```diff
@@ -579,7 +579,7 @@ static PyObject *FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DW
 	}
 
 	if (flags == EvtFormatMessageKeyword) {
-		buf[returned_size * sizeof(WCHAR)] = L'\0';
+		buf[returned_size] = L'\0';
```

Comment:
This still seems wrong to me, but the MS docs aren't clear, and in some cases, simply wrong.

`returned_size` being characters means this is still going to be sticking a null in the middle of the string. Let's say `returned_size` comes back as 10 - this means we need a buffer for 10 wide chars, so 20 bytes. We'll correctly allocate the 20 bytes, but here we will stick the \0 10 *bytes* into that buffer. Up at line 561 we've ensured that the buffer is large enough to take this extra \0, and we've done this before multiplying by `sizeof(WCHAR)` - so this looks perfectly correct to me as it stands, and your change will stick a \0 in the middle of the buffer and may not leave a double-null terminated string.

(https://docs.microsoft.com/en-us/windows/win32/wes/formatting-event-messages is about the only example I can find trying to use this, but it's clearly wrong in that it does:

```
            if ((EvtFormatMessageKeyword == FormatId))
                pBuffer[dwBufferSize-1] = L'\0';
            else
                dwBufferSize = dwBufferUsed;
```
when `pBuffer` is null and `dwBufferSize` is zero!

The only other possibility I can see is that `returned_size` does not count the "normal" trailing \0 character - ie, if the returned string was going to be 'X', then `returned_size` might return 1 - meaning we need to allocate space for *2* chars (ie, for `X\0`) and *3* chars for `EvtFormatMessageKeyword` (so we end up with `X\0\0` in that case). Does you crash go away if line 560 changes to `allocated_size = returned_size + 1;`?

There aren't any samples which use `EvtFormatMessageKeyword` - can you maybe share one?

Reply:
>returned_size being characters means this is still going to be sticking a null in the middle of the string. Let's say returned_size comes back as 10 - this means we need a buffer for 10 wide chars, so 20 bytes. We'll correctly allocate the 20 bytes, but here we will stick the \0 10 bytes into that buffer.

But what about that the type of the `buf` is `LPWSTR`? Does that mean that when we do `buf[returned_size]`, we access 10th two-byte long WCHAR **element**, not byte, and this element should be located in the 19th and 20th byte respectively?

Reply:
doh! Yes, you are correct.